### DISCUSSION TITLE
[A-L] (18 maggio 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#97] [Esercitazioni A-L]
![immagine](https://github.com/user-attachments/assets/f1561f62-c90f-4082-bde1-ae1a07078857)
![immagine](https://github.com/user-attachments/assets/ef667151-07b9-4ce6-b876-6be31b91ae9c)

**Comment by cornflexxx** [comment]
$1)$
-  $d(X,X) = 0\rightarrow$ la stringa $X$ è uguale a se stessa per ogni $X[i]$ e quindi non necessita di modifiche. 
- $d(X,Y) = d(Y,X) \rightarrow$ Supponiamo di voler trasformare la stringa $X$ in $Y$ quindi possiamo applicare  le operazioni che ci permettono di effettuare il minor numero di modifiche per arrivare ad $Y$. Procedendo a ritroso, una volta giunti ad $Y$ possiamo tornare ad $X$ con lo stesso numero minimo di operazioni questa volta invertendole col seguente criterio: $sostituzione(i,j)\to sostituzione(j,i), cancellazione(k)\to inserimento(k,z), inserimento(k,z)\to cancellazione(k)$. 
In questo modo otteniamo $X$ a partire da $Y$  con stesso numero di modifiche e quindi $d(X,Y) = d(Y,X)$
- $d(X,Y) \le d(X,Z) + d(Z,Y)\rightarrow$   supponiamo che $d(X,Y) \gt d(X,Z) + d(Z,Y)$ implicando che  esiste una stringa intermedia $Z$ tra $X$ e $Y$, tale che passando per questa riusciamo a trasformare $X$ in $Y$ con un numero di modifiche minori a $d(X,Y)$, contraddicendo il fatto che $d(X,Y)$ sia  minima. Quindi abbiamo due alternative:
la stringa intermedia $Z$ può essere ottenuta dopo $i$ modifiche ad $X$ per arrivare ad $Y$, quindi fa parte del "percorso minimo" $X\to Z\to Y$ e quindi $d(X,Y) = d(X,Z) + d(Z,Y)$.
Oppure la stringa $Z$ non fa parte del percorso minimo di modifiche e quindi "allunga" tale percorso rendendo ottendendo la seguente disuguaglianza $d(X,Y) \le d(X,Z) +d(Z,Y)$.

Notiamo che $d(X,X) = 0 \le d(X,Y) + d(Y,X) $, supponendo  $Y\neq X$ e quindi $d(X,Y)>0$

$2)$ 
Consideriamo la matrice $T_{m\times n}$ dove $T[i,j] \to$ numero minimo di modifiche necessarie per trasformare $i$ caratteri di  $X$ in $j$ di $Y$

$$T[i,j]= \left\\{ \begin{array}{cl}
j &  i = 0\\
i &  j = 0\\
T[i-1,j-1]& X[i] = Y[j]\\
\min(T[i-1,j-1], T[i-1,j], T[i,j-1])+1 & altrimenti
\end{array} \right.$$

Se $X[i] \neq Y[j]$ posso decidere di effettuare una delle 3 operazioni :  

- Eliminare $X[i]$   e di conseguenza  dover convertire i restanti $i-1$  caratteri in $j$
- Eliminare $Y[j]$ e di conseguenza dover convertire i restnati $i$ caratteri in $j-1$
- Sostituire  e di conseguenza lavorare su i restanti $i-1$ da convertire in $j-1$

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/3407f761-9236-43c3-8ccf-52c79f0eae18)
